# module load MuSE/1.0rc


# Command: call    Call somatic point mutations.
#                  Raw output file will be created. If WGS
#                  data is used, it is recommended to split
#                  the WGS data into small blocks (<50Mb).
#          sump    Generate final calls in VCF format.
#                  Besides PASS calls, tier-based calls are
#                  reported using dynamic cutoff searching
#                  approach. If there are multiple raw output
#                  files from WGS data, please have them
#                  concatenated first.

# ./MuSE call –O Output.Prefix –f Reference.Genome Tumor.bam Matched.Normal.bam
# ./MuSE sump -I Output.Prefix.MuSE.txt -G –O Output.Prefix.vcf –D dbsnp.vcf.gz

# Usage:   MuSE call [options] tumor.bam matched_normal.bam
# Options:
#          -f FILE    faidx indexed reference sequence file
#          -r STR     single region (chr:pos-pos) where somatic
#                     mutations are called
#          -l FILE    list of regions (chr:pos-pos or BED), one
#                     region per line
#          -O STR     output file name (suffix '.MuSE.txt' is
#                     automatically added)
muse <- function(
      trk, 
      samplename,
      ref_fasta = opts_flow$get('ref_fasta'),
      split_by_chr = TRUE,
      java_mem_str = opts_flow$envir$java_mem_str,
      gatk4_exe = opts_flow$envir$gatk4_exe,
      mutect1_germline_vcf = opts_flow$envir$mutect1_germline_vcf

){

  check_args()
  flog.debug(glue("Generating Muse flowmat for sample: {samplename}"))

  trk <- metadata_for_mutect1(trk)
  expect_columns(trk, "outprefix")
  if (variant_calling_mode == "matched") {
        expect_columns(trk, "outprefix_paired")
    }

  trk_tum <- filter(trk, normal == "NO")
  trk_norm <- filter(trk, normal == "YES")

  i <- 1
  out = lapply(1:nrow(trk_tum), function(i){

    tumor_bam <- trk_tum$bam[i]
    normal_bam <- trk_norm$bam[1] # assuming a SINGLE normal!!
    tumor_name <- trk_tum$name[i]
    normal_name <- trk_norm$name[1]
    outprefix <- trk_tum$outprefix_paired[i]

    bamset = bam_set(bam = tumor_bam, outprefix = outprefix,
            ref_fasta = ref_fasta, split_by = "chr")

    # create split cmds
    intervals = bamset$intervals %>% paste0("-r ", .)
    outprefix_chr = bamset$outprefix_chr
    muse_calls <- paste0(bamset$outprefix_chr, ".MuSE.txt")
    muse_vcfs <- paste0(bamset$outprefix_chr, ".vcf")
    cmd_calls = glue("module load MuSE/1.0rc;MuSE call -O {outprefix_chr} -f {ref_fasta} {tumor_bam} {normal_bam} {intervals}")
    # ./MuSE sump -I Output.Prefix.MuSE.txt -G –O Output.Prefix.vcf –D dbsnp.vcf.gz
    # -E: whole exome
    cmd_sumps = glue("MuSE sump -I {muse_calls} -E -O {muse_vcfs} -D {mutect1_germline_vcf}")

    # merge VCF (using GATK)
    muse_vcfs_i = paste0(muse_vcfs, collapse = " -I ")
    muse_vcf = glue("{outprefix}.vcf.gz")
    cmd_mergevcf = glue(
          "{gatk4_exe} --java-options {java_mem_str} GatherVcfs -I {muse_vcfs_i} -O {muse_vcf}; ",
          "{gatk4_exe} IndexFeatureFile -I {muse_vcf}")

    # call it merge, since one depends on the other
    cmds <- list(muse.splt = glue("{cmd_calls};{cmd_sumps}"),
                 muse.mrg = cmd_mergevcf)
    flowmat = to_flowmat(cmds, samplename = samplename) %>%
        mutate(cmd = as.character(cmd))
    outfiles = list(muse_vcf = muse_vcf)

    list(cmds = cmds, flowmat = flowmat, outfiles = outfiles)
  })
  out = transpose(out)
  flowmat = out$flowmat %>% bind_rows()
  outfiles = out$outfiles

  return(list(flowmat = flowmat, outfiles = outfiles))
}

# Usage:   MuSE sump [options]
# Options:
#          -I FILE    single input file generated by 'MuSE call'
#          -G         input generated from whole genome sequencing data
#          -E         input generated from whole exome sequencing data
#          -O STR     output file name (VCF format)
#          -D FILE    dbSNP vcf file that should be bgzip compressed,
#                     tabix indexed and based on the same reference
#                     genome used in 'MuSE call'