


optitype <- function(samplename, 
                     fqs,
                     seq_sample_id,
                     optitype_sif = opts_flow$get("optitype_sif"),
                     sample_type = c("dna", "rna"),
                     funr_exe = opts_flow$get("R_funr_exe"),
                     paired_end = T,
                     odir = "optitype"
                     
                     
                     ){
  
  fq1=fqs$fq1;fq2=fqs$fq2;fq3=fqs$fq3
  sample_type = match.arg(sample_type)
  
  check_args()
  
  # https://github.com/FRED-2/OptiType
  # beta value for heterozyhozyty 0.009 (change with caution)
  # --enumerate: top 1 solution
  # --input FQ [FQ], -i FQ [FQ]
  # .fastq file(s) (fished or raw) or .bam files stored
  # for re-use, generated by an earlier OptiType run. One
  # file: single-end mode, two files: paired-end mode.
  if(sample_type == "dna"){
    if(paired_end)
      cmd_optitype = glue("singularity exec {optitype_sif} OptiTypePipeline.py -i {fq1} {fq2} --dna --beta 0.009 --enumerate 1 --verbose --outdir optitype")
    else
      cmd_optitype = glue("singularity exec {optitype_sif} OptiTypePipeline.py -i {fq3} --dna --beta 0.009 --enumerate 1 --verbose --outdir optitype")
  }else if(sample_type == "rna"){
    if(paired_end)
      cmd_optitype = glue("singularity exec {optitype_sif} OptiTypePipeline.py -i {fq1} {fq2} --rna --beta 0.009 --enumerate 1 --verbose --outdir optitype")
    else
      cmd_optitype = glue("singularity exec {optitype_sif} OptiTypePipeline.py -i {fq3} --dna --beta 0.009 --enumerate 1 --verbose --outdir optitype")
  }
  
  cmd_optitype = glue("mkdir {odir};{cmd_optitype}")
  
  # seq_sample_id = seq_sample_id_wex_n
  optitype_pvac_mhc1 = glue("{odir}/{seq_sample_id}_optitype_pvac_mhc1.txt")
  cmd_optitype_pvac = glue("{funr_exe} my.ultraseq::to_pvacseq.optitype path={odir} outfile={optitype_pvac_mhc1}")
  # to_pvacseq.optitype(path, optitype_pvac)

  cmds = list(optitype = c(cmd_optitype, cmd_optitype_pvac))
  flowmat = to_flowmat(cmds, samplename)  
  
  outfiles = list(optitype_pvac_mhc1 = optitype_pvac_mhc1)
  
  list(flowmat = flowmat, outfiles = outfiles)
}


#' to_pvacseq.optitype
#' 
#' convert optitype for pvacseq
#'
#' @param path 
#' @param outfile outfile
#'
#' @export
to_pvacseq.optitype <- function(path, outfile){
  
  # path = "~/.rsrch3/scratch/iacs/sseth/flows/SS/tnbc/ms51_wex/neoantigen/runs/185_003/tmp/optitype"
  # fl = list.files(path, pattern = "result.tsv", recursive = T) %>% tail(1)
  fl = fs::dir_ls(path, glob = "*/*_result.tsv", recurse = T) %>% tail(1)
  df_opti = data.table::fread(fl, data.table = F)
  hla_alleles =   dplyr::select(df_opti, A1:C2) %>% unlist() %>% 
    paste0("HLA-", ., collapse = ",") %>% 
    mutate(allele1 = hla_trim_d4(allele1),
           allele2 = hla_trim_d4(allele2))
  
  write(hla_alleles, outfile)
  outfile

}