## in-house



star.2.4.0j <- function(fq1, fq2,
  samplename = "samp1",
  out_path, outprefix,
  star_exe = opts_flow$get("star_exe"),
  ref_fasta = opts_flow$get("ref_fasta"),
  ref_star = opts_flow$get("ref_star"),
  star_opts = opts_flow$get("star_opts"),
  star_cpu = opts_flow$get("star_cpu")){

  .Deprecated("2.4.2a")

  ## detect gz and change options
  if(grepl("\\.gz$", fq1)){
    star_opts = paste0(star_opts, " --readFilesCommand zcat")
  }

  if(missing(fq2))
    fq = fq1
  else
    fq = c(fq1, fq2)

  chkfq <- chk_fq(fqs1 = fq1, fqs2 = fq2)

  if(missing(outprefix))
    outprefix = gsub(chkfq$ext, "", basename(fq1))

  prefix1 <- paste0(outprefix,"initial",sep='') ## prefix
  prefix2 <- paste0(outprefix,"star", sep='') ## prefix
  ref2_star <- paste0(outprefix, 'genomedir')

  ## -- files generated by star
  sjtab <- paste0(prefix1, "_SJ.out.tab")
  outsam <- paste0(prefix1, "Aligned.out.sam")
  sjtab2 <- paste0(prefix2, "_SJ.out.tab")
  outsam2 <- paste0(prefix2, "Aligned.out.sam")

  cmd_align1 <- sprintf("%s --genomeDir %s --readFilesIn %s --runThreadN %s --outFileNamePrefix %s_ %s",
    star_exe, ref_star, fq, star_cpu, prefix1, star_opts)

  cmd_index = sprintf("mkdir %s; %s --runMode genomeGenerate --genomeFastaFiles %s --sjdbFileChrStartEnd %s --genomeDir %s %s",
    ref2_star, star_exe, ref_fasta, sjtab, ref2_star, star_opts)

  cmd_align2 <- sprintf("%s --genomeDir %s --readFilesIn %s --runThreadN %s --outFileNamePrefix %s_ %s",
    star_exe, ref2_star, fq, star_cpu, prefix2, star_opts)

  cmds = list(align1 = cmd_align1, index = cmd_index, align2 = cmd_align2)

  flowmat = to_flowmat(cmds, samplename)
  return(list(flowmat = flowmat, outfiles = c(sjtab2, outsam2)) )
}





## input from ion-torrent is preprocessed:
## @PG     ID:bc   PN:BaseCaller   VN:4.2-14/88431 CL:BaseCaller --barcode-filter 0.01 --barcode-filter-minreads 10 --keypass-filter on --phasing-residual-filter=2.0 --num-unfiltered 1000 --barcode-filter-postpone 1 --barcode-mode 1 --barcode-cutoff 2 --input-dir=sigproc_results --librarykey=TCAG --tfkey=ATCG --run-id=V4IOG --output-dir=basecaller_results --block-col-offset 0 --block-row-offset 7992 --datasets=basecaller_results/datasets_pipeline.json --trim-adapter ATCACCGACTGCCCATAGAGAGGCTGAGAC

#' @rdname star
#' @usage star.2.4.2a
star.2.4.2a <- function(fq1, fq2, bam,
  samplename = "samp1",
  out_path,
  outprefix,
  ref_fasta = opts_flow$get("ref_fasta"),
  ref_star = opts_flow$get("ref_star"),
  star_exe = opts_flow$get("star_exe"),
  star_opts = opts_flow$get("star_opts"),
  star_cpu = opts_flow$get("star_cpu")
  ){


  assertthat::assert_that(is.character(ref_fasta))
  assertthat::assert_that(is.character(ref_star))
  assertthat::assert_that(is.character(star_exe))
  assertthat::assert_that(is.character(star_opts))
  assertthat::assert_that(!is.null(star_cpu)) ## not numeric, might be {{{CPU}}}


  if(!missing(fq1)){
    if(!missing(bam))
      stop("Supply either fq OR bam as input, not both. Refer to STAR's manual for more details.")

    ## detect gz and change options
    if(grepl("\\.gz$", fq1)){
      star_opts = paste0(star_opts, " --readFilesCommand zcat")
    }

    if(missing(fq2))
      fq = fq1
    else
      fq = c(fq1, fq2)

    chkfq <- chk_fq(fqs1 = fq1, fqs2 = fq2)

    if(missing(outprefix)) ## guess outprefix
      outprefix = gsub(chkfq$ext, "", basename(fq1))

    cmd_star <- sprintf("%s  --twopassMode Basic --genomeDir %s --readFilesIn %s --runThreadN %s --outFileNamePrefix %s_ %s",
      star_exe, ref_star, fq, star_cpu, outprefix, star_opts)
  }


  if(!missing(bam)){

    if(missing(outprefix)) ## guess outprefix
      outprefix = gsub(".bam", "", basename(bam))

    ## notes:
    ## EXITING because of fatal INPUT ERROR: at the moment --runMode inputFromBAM only works with --outWigType bedGraph OR --bamRemoveDuplicatesType Identical
    cmd_star <- sprintf("%s  --twopassMode Basic --genomeDir %s --inputBAMfile %s --runMode inputAlignmentsFromBAM --outWigType bedGraph --runThreadN %s --outFileNamePrefix %s_ %s",
      star_exe, ref_star, bam, star_cpu, outprefix, star_opts)
  }

  if(missing(bam) & missing(fq1))
    stop("Both inputs, bam and fastq are missing !")

  flowmat = to_flowmat(list(star = cmd_star), samplename)
  return(list(flowmat = flowmat, outfiles = c(outprefix)) )


}

# https://github.com/mskcc/RNAseqDB/blob/master/calc-expression.pl
# STAR parameters: common (options used here are similar as https://github.com/ENCODE-DCC/long-rna-seq-pipeline/blob/master/dnanexus/align-star-pe/resources/usr/bin/lrna_align_star_pe.sh)
# my $star_para_comm = "--genomeDir $index_star --readFilesIn $fastq1 $fastq2  --outFilterType BySJout --outSAMattributes NH HI AS NM MD  --outFilterMultimapNmax 20   --outFilterMismatchNmax 999  --outFilterMismatchNoverLmax 0.04   --alignIntronMin 20   --alignIntronMax 1000000   --alignMatesGapMax 1000000 --alignSJoverhangMin 8  --alignSJDBoverhangMin 1";
# my $star_bam = "--outSAMtype BAM Unsorted --quantMode TranscriptomeSAM"
# STAR parameters: metadata
# my $star_meta = '--outSAMheaderHD @HD VN:1.4 SO:coordinate';
# `$star_bin $star_para_comm $read_fq_cmd $star_run $star_bam $star_meta`;
# index
# `samtools sort -o Aligned.sortedByCoord.out.bam -T Aligned.sortedByCoord.out.bam.0 -O bam Aligned.out.bam`;
# `samtools index Aligned.sortedByCoord.out.bam`;
# Run RSEM on BAM file
# my $rsem_para_comm = "--bam --estimate-rspd  --calc-ci --no-bam-output --seed $rsem_rnd_seed";
# 
# # RSEM parameters: run-time, number of threads and RAM in MB
# my $rsem_run = "-p $thread_n --ci-memory 30000";
# 
# # RSEM parameters: data type dependent
# my $gencode_bed = $gencode;
# $gencode_bed =~ s/gtf/bed/;
# 
# my $data_type = GetStrandness( $gencode_bed, 'Aligned.sortedByCoord.out.bam' );
# my $rsem_para_type;
# 
# switch ($data_type) {
#   case 'str_SE'   { $rsem_para_type = "--forward-prob 0" }
#   case 'str_PE'   { $rsem_para_type = "--paired-end --forward-prob 0" }
#   case 'unstr_SE' { $rsem_para_type = "" }
#   case 'unstr_PE' { $rsem_para_type = "--paired-end" }
#   else		    { die "ERROR: unknown data type\n" }
# }
# 
# # RSEM command
# print "Calculate expression using RSEM...\n";
# print "\n$rsem_dir/rsem-calculate-expression $rsem_para_comm $rsem_run $rsem_para_type Aligned.toTranscriptome.out.bam $index_rsem Quant >& Log.rsem\n";
# (-e 'Quant.genes.results') or `$rsem_dir/rsem-calculate-expression $rsem_para_comm $rsem_run $rsem_para_type Aligned.toTranscriptome.out.bam $index_rsem Quant >& Log.rsem`;
# 

#' @rdname star
#' @usage star.2.4.2a
star.2.4.2a <- function(fq1, fq2, bam,
                        samplename = "samp1",
                        out_path,
                        outprefix,
                        ref_fasta = opts_flow$get("ref_fasta"),
                        ref_star = opts_flow$get("ref_star"),
                        star_exe = opts_flow$get("star_exe"),
                        star_opts = opts_flow$get("star_opts"),
                        star_cpu = opts_flow$get("star_cpu")
){
  
  
  assertthat::assert_that(is.character(ref_fasta))
  assertthat::assert_that(is.character(ref_star))
  assertthat::assert_that(is.character(star_exe))
  assertthat::assert_that(is.character(star_opts))
  assertthat::assert_that(!is.null(star_cpu)) ## not numeric, might be {{{CPU}}}
  
  
  if(!missing(fq1)){
    if(!missing(bam))
      stop("Supply either fq OR bam as input, not both. Refer to STAR's manual for more details.")
    
    ## detect gz and change options
    if(grepl("\\.gz$", fq1)){
      star_opts = paste0(star_opts, " --readFilesCommand zcat")
    }
    
    if(missing(fq2))
      fq = fq1
    else
      fq = c(fq1, fq2)
    
    chkfq <- chk_fq(fqs1 = fq1, fqs2 = fq2)
    
    if(missing(outprefix)) ## guess outprefix
      outprefix = gsub(chkfq$ext, "", basename(fq1))
    
    cmd_star <- sprintf("%s  --twopassMode Basic --genomeDir %s --readFilesIn %s --runThreadN %s --outFileNamePrefix %s_ %s",
                        star_exe, ref_star, fq, star_cpu, outprefix, star_opts)
  }
  
  
  if(!missing(bam)){
    
    if(missing(outprefix)) ## guess outprefix
      outprefix = gsub(".bam", "", basename(bam))
    
    ## notes:
    ## EXITING because of fatal INPUT ERROR: at the moment --runMode inputFromBAM only works with --outWigType bedGraph OR --bamRemoveDuplicatesType Identical
    cmd_star <- sprintf("%s  --twopassMode Basic --genomeDir %s --inputBAMfile %s --runMode inputAlignmentsFromBAM --outWigType bedGraph --runThreadN %s --outFileNamePrefix %s_ %s",
                        star_exe, ref_star, bam, star_cpu, outprefix, star_opts)
  }
  
  if(missing(bam) & missing(fq1))
    stop("Both inputs, bam and fastq are missing !")
  
  flowmat = to_flowmat(list(star = cmd_star), samplename)
  return(list(flowmat = flowmat, outfiles = c(outprefix)) )
  
  
}



# STAR DOC

#' @title A flowr wrapper for STAR aligner
#'
#' @description For more details refer to \href{https://github.com/alexdobin/STAR/raw/master/doc/STARmanual.pdf}{STAR's manual}.
#' Starting 2.4.2a, star supports a twoPass mode, thus simplifying this pipeline.
#'
#' @usage
#' star(fq1, fq2, bam,
#'  samplename = "samp1", out_path, outprefix,
#'  ref_fasta = opts_flow$get("ref_fasta"), ref_star = opts_flow$get("ref_star"),
#'  star_exe = opts_flow$get("star_exe"), star_opts = opts_flow$get("star_opts"),
#'  star_cpu = opts_flow$get("star_cpu"))
#'
#' @param fq1 single fastq file
#' @param fq2 second read fastq file
#' @param bam alternatively use of bam as input, instead of fqs
#' @param samplename name of the sample
#' @param out_path output path
#' @param outprefix output prefix
#' @param star_exe path to star executable
#' @param ref_fasta path to reference genome in fasta format
#' @param ref_star path to star's reference index
#' @param star_opts additional options passed onto star. This is a simple character vector appended to the commandline.
#' @param star_cpu a integer specifying the number of cores to be used
#' @export
star <- star.2.4.2a



## would get from flow_def

#' star_index
#' @description a helper function to create a index, incase it does not exits
star_index <- function(
  ref_star = opts_flow$get("ref_star"),
  star_exe = opts_flow$get("star_exe"),
  ref_fasta = opts_flow$get("ref_fasta"),
  star_cpu = opts_flow$get("star_cpu"),
  ref_gtf = opts_flow$get("ref_gtf")){

  sprintf("%s --runMode genomeGenerate --genomeDir %s --genomeFastaFiles %s --runThreadN %s",
    star_exe, ref_star,  ref_fasta, star_cpu)

}


if(FALSE){
  ## get a star index
  # 
  # opts_flow$set(
  #   star_opts = "--sjdbOverhang 75",
  #   star_cpu = "{{{CPU}}}"
  # )

  #debug(star_pipe)
  ref_fasta = '/scratch/rists/hpcapps/reference/human/broad_hg19/fastas/Homo_sapiens_assembly19.fasta'
  ref_star = '/scratch/rists/hpcapps/reference/human/broad_hg19/indexes/STAR/2.3.0e'

  fq1 = "/rsrch2/iacs/ngs_runs/1411_sarcomatoid/plugin_out/downloads/1711-N.RNA_Barcode_None_001.user_PRO-125-Sarcomatoid_RNA_Seq_1711-N.fastq"
  out = star(fq1, star_cpu = 8, ref_fasta = ref_fasta, ref_star = ref_star)


}


## example star pipeline for ion-torrent
if(FALSE){

  undebug(whisker:::parseTemplate)
  debug(params:::parse_conf)

  load_opts(fetch_conf("ngsflows_mda.conf"))
  opts_flow$get()

  cmdindex = star_index(ref_gtf = opts_flow$get("ref_hg19_gtf"))


  bam = "/rsrch2/iacs/ngs_runs/1411_sarcomatoid/bams-2015.07.24/RNA_Barcode_None_user_PRO-64-Sarcomatoid_RNA_Seq_768-T_190.bam"
  out = star(bam, ref_fasta = ref_fasta, ref_star = ref_star, star_opts = "--sjdbOverhang 100")
  out

}
